src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -324,6 +325,94 @@ library GPv2Encoding {
         }
     }
 
+    /// @dev Decodes an interaction from calldata into memory.
+    ///
+    /// An encoded interaction has three components: the target address, the data
+    /// length, and the actual interaction data of variable size.
+    ///
+    /// ```
+    /// struct EncodedInteraction {
+    ///     address target;
+    ///     uint24 dataLength;
+    ///     bytes callData;
+    /// }
+    /// ```
+    ///
+    /// All entries are tightly packed together in this order in the encoded
+    /// calldata. Example:
+    ///
+    /// input:    0x73c14081446bd1e4eb165250e826e80c5a523783000010000102030405060708090a0b0c0d0e0f
+    /// decoding:   [...............target.................][leng][............data..............]
+    /// stride:                                          20     3    (defined in length field) 16
+    ///
+    /// This function enforces that the encoded data stores enough bytes to
+    /// cover the full length of the decoded interaction.
+    ///
+    /// The size of `dataLength` limits the maximum calldata that can be used in
+    /// an interaction. Based on the current rules of the Ethereum protocol,
+    /// this length is enough to include any valid transaction: an extra
+    /// calldata byte costs at least 4 gas, and the maximum gas spent in a block
+    /// is 12.5M. This gives an upper bound on the calldata that can be included
+    /// in a block of
+    ///   3.125.000 < 16.777.216 = 2**(3*8) .
```

Comment:
Very good reasoning! At first I thought a little weird to use a `u24` as its a non-conventional type, but I think this reasoning is very good and explains exactly why this size is used.

One related question is do you think its worth making this a `u32` incase they increase the block gas limit by a factor of 6 for example (it has increased in the past) or reduce calldata costs even further (also reduced in the past)?

Reply:
Especially given the content of Tim Roughgarden's recent publication and EIP where he suggests that the block gas limit can vary up to double the target value of 12.5M.

http://timroughgarden.org/papers/eip1559.pdf

Reply:
The upper bound is very generous in our case I think: In practice for us the bottleneck is memory. The calldata is copied to memory for the full interaction, and this cannot be avoided as long as we `CALL` the contract. An `MSTORE` for the `w`-th word in memory costs `3w + FLOOR(w²/512)` gas currently. So just storing the last byte of a calldata of length 2²⁴ costs ≥538M, and storing everything costs ≥10¹⁴ gas.
In practice, already in my test I had to limit myself to a calldata of length `12000`. In this case just the cost of writing the result to memory (not counting e.g. reading calldata) appears to be `≥244k` gas already.

All these numbers might change in the future, but I'd be surprised if they were decreased by so many orders of magnitude. So if anything I'd consider dropping another bit, since the cost of writing 2¹⁶ bytes to memory is ≥11M. But in this case I prefer to be a bit future-proof and waste a byte per order.

<details>
<summary>
Math
</summary>

```
> m = (x) => 3*x/32+Math.floor(x*x/32/32/512)
> writeToMem = (x) => [...Array(Math.floor(x/32))].map((_, i) => m(32*i)).reduce((a,b)=> a+b, 0)
> m(2**24)
538443776
> Math.log(writeToMem(2**24))/Math.log(10)
13.974221634762877
> writeToMem(12000)
244404
> writeToMem(2**16)
11875792
```
</details>